src/Microsoft.ML.TimeSeries/SrCnnTransformBase.cs

```diff
@@ -0,0 +1,318 @@
+            AvergingWindowSize = averagingWindowSize;
+            JudgementWindowSize = judgementWindowSize;
+            AlertThreshold = alertThreshold;
+
+            OutputLength = 3;
+        }
+
+        private protected SrCnnTransformBase(IHostEnvironment env, ModelLoadContext ctx, string name)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(name), ctx)
+        {
+            OutputLength = 3;
+
+            byte temp;
+            temp = ctx.Reader.ReadByte();
+            BackAddWindowSize = (int)temp;
+            Host.CheckDecode(BackAddWindowSize > 0);
+
+            temp = ctx.Reader.ReadByte();
+            LookaheadWindowSize = (int)temp;
+            Host.CheckDecode(LookaheadWindowSize > 0);
+
+            temp = ctx.Reader.ReadByte();
+            AvergingWindowSize = (int)temp;
+            Host.CheckDecode(AvergingWindowSize > 0);
+
+            temp = ctx.Reader.ReadByte();
+            JudgementWindowSize = (int)temp;
+            Host.CheckDecode(JudgementWindowSize > 0);
+
+            AlertThreshold = ctx.Reader.ReadDouble();
+            Host.CheckDecode(AlertThreshold >= 0 && AlertThreshold <= 1);
+        }
+
+        private protected SrCnnTransformBase(SrCnnArgumentBase args, string name, IHostEnvironment env)
+            : this(args.WindowSize, args.InitialWindowSize, args.Source, args.Name,
+                  name, env, args.BackAddWindowSize, args.LookaheadWindowSize, args.AvergingWindowSize, args.JudgementWindowSize, args.Threshold)
+        {
+        }
+
+        private protected override void SaveModel(ModelSaveContext ctx)
+        {
+            Host.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel();
+
+            Host.Assert(WindowSize > 0);
+            Host.Assert(InitialWindowSize == WindowSize);
+            Host.Assert(BackAddWindowSize > 0);
+            Host.Assert(LookaheadWindowSize > 0);
+            Host.Assert(AvergingWindowSize > 0);
+            Host.Assert(JudgementWindowSize > 0);
+            Host.Assert(AlertThreshold >= 0 && AlertThreshold <= 1);
+
+            base.SaveModel(ctx);
+            ctx.Writer.Write((byte)BackAddWindowSize);
+            ctx.Writer.Write((byte)LookaheadWindowSize);
+            ctx.Writer.Write((byte)AvergingWindowSize);
+            ctx.Writer.Write((byte)JudgementWindowSize);
+            ctx.Writer.Write(AlertThreshold);
+        }
+
+        internal override IStatefulRowMapper MakeRowMapper(DataViewSchema schema) => new Mapper(Host, this, schema);
+
+        internal sealed class Mapper : IStatefulRowMapper
+        {
+            private readonly IHost _host;
+            private readonly SrCnnTransformBase<TInput, TState> _parent;
+            private readonly DataViewSchema _parentSchema;
+            private readonly int _inputColumnIndex;
+            private readonly VBuffer<ReadOnlyMemory<Char>> _slotNames;
+            private SrCnnStateBase State { get; set; }
```

Comment:
This looks like a private field. Not sure if you wan to do `private SrCnnStateBase _state`

Reply:
_state does not match PascalCased